integration/test/ParseEventuallyQueueTest.js

```diff
@@ -0,0 +1,238 @@
+    await Parse.EventuallyQueue.destroy(results[2]);
+
+    length = await Parse.EventuallyQueue.length();
+    assert.strictEqual(length, 3);
+
+    await Parse.EventuallyQueue.sendQueue();
+    const objects = await query.find();
+    assert.strictEqual(objects.length, 0);
+  });
+
+  it('can queue destroy for object that does not exist', async () => {
+    const object = new TestObject({ test: 'test' });
+    await object.save();
+    await object.destroy();
+    await Parse.EventuallyQueue.destroy(object);
+    await Parse.EventuallyQueue.sendQueue();
+
+    const length = await Parse.EventuallyQueue.length();
+    assert.strictEqual(length, 0);
+  });
+
+  it('can queue destroy then save', async () => {
+    const object = new TestObject({ hash: 'test' });
+    await Parse.EventuallyQueue.destroy(object);
+    await Parse.EventuallyQueue.save(object);
+    await Parse.EventuallyQueue.sendQueue();
+
+    const query = new Parse.Query(TestObject);
+    query.equalTo('hash', 'test');
+    const results = await query.find();
+    assert.strictEqual(results.length, 1);
+
+    const length = await Parse.EventuallyQueue.length();
+    assert.strictEqual(length, 0);
+  });
+
+  it('can queue unsaved object with hash', async () => {
+    const hash = 'secret';
+    const object = new TestObject({ test: 'test' });
+    object.set('hash', hash);
+    await Parse.EventuallyQueue.save(object);
+    await Parse.EventuallyQueue.sendQueue();
+
+    const query = new Parse.Query(TestObject);
+    query.equalTo('hash', hash);
+    const results = await query.find();
+    assert.strictEqual(results.length, 1);
+  });
+
+  it('can queue saved object and unsaved with hash', async () => {
+    const hash = 'ransom+salt';
+    const object = new TestObject({ test: 'test' });
+    object.set('hash', hash);
+    await Parse.EventuallyQueue.save(object);
+    await Parse.EventuallyQueue.sendQueue();
+
+    let query = new Parse.Query(TestObject);
+    query.equalTo('hash', hash);
+    const results = await query.find();
+    assert.strictEqual(results.length, 1);
+
+    const unsaved = new TestObject({ hash, foo: 'bar' });
+    await Parse.EventuallyQueue.save(unsaved);
+    await Parse.EventuallyQueue.sendQueue();
+
+    query = new Parse.Query(TestObject);
+    query.equalTo('hash', hash);
+    const hashes = await query.find();
+    assert.strictEqual(hashes.length, 1);
+    assert.strictEqual(hashes[0].get('foo'), 'bar');
+  });
+
+  it('can queue same object but override undefined fields', async () => {
+    const object = new Parse.Object('TestObject');
+    object.set('foo', 'bar');
+    object.set('test', '1234');
+    await Parse.EventuallyQueue.save(object);
+
+    object.set('foo', undefined);
+    await Parse.EventuallyQueue.save(object);
+
+    const length = await Parse.EventuallyQueue.length();
+    assert.strictEqual(length, 1);
+
+    const queue = await Parse.EventuallyQueue.getQueue();
+    assert.strictEqual(queue[0].object.foo, 'bar');
```

Comment:
This should actually be `undefined`, according to this test:

```js
const o = new Parse.Object('Tmp');
o.set('key', 'override1');
o.set('key', undefined);
await o.save(null, {useMasterKey: true});
```
Creates an object with `key: undefined`.

Reply:
Ah makes sense

Reply:
Another optimization could be save-delete:
1. Add save new obj operation to queue.
2. Add delete new object operation to queue.
3. Optimize: delete overrides save and therefore removes both entries from queue.

Although, still I am not sure these optimizations would be intuitive or wanted by everyone? Saving an object can trigger hooks in cloud code, deleting an object as well. So someone may want to save, then delete an object and expect triggers to be executed in between. If the two ops are optimized in the queue, they never reach the server. Therefore these optimizations maybe should be optional? Not sure.

Reply:
Maybe also a test case that ensures that no optimization takes places if two operations are not adjacent to each other?

Reply:
@dplewis I think your test case is right after all:

a)
```js
const o = new Parse.Object("Temp");
o.set('key', 'override');
await o.save(null, {useMasterKey: true});
o.set('key', undefined);
await o.save(null, {useMasterKey: true}); // results in object with `key: override`
```

b)
```js
const o = new Parse.Object("Temp");
o.set('key', 'override');
o.set('key', undefined);
await o.save(null, {useMasterKey: true}); // results in object with `key: undefined`
```

It makes sense considering that `undefined` is currently interpreted as "no value set" and therefore ignored when saving an object.

Reply:
Thanks for looking into this. I think this is ready to merge for now